### PR TITLE
Correction : InvalidParameterException and metadata

### PR DESCRIPTION
**Pull request by FlorianBouron**
Adding some changes about:
1. InvalidParameterException :  removing the error with InvalidParameterException issue #49

2. meta data : When a log is created without "meta" data, it doesn't appear in the log of Cloudwatch.

Example of log with meta data : 
- BEFORE : "level - message - {meta}"
- NOW : "level - message - {meta}"

Example of log without meta data : 
- BEFORE : "level - message - {meta}"
- NOW : "level - message"

3. Improvement : We don't reach Cloudwatch API if we don't have new log (less API request).

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -17,7 +17,12 @@ var WinstonCloudWatch = function(options) {
   var awsSecretKey = options.awsSecretKey;
   var awsRegion = options.awsRegion;
   var messageFormatter = options.messageFormatter ? options.messageFormatter : function(log) {
-    return [ log.level, log.msg, stringify(log.meta) ].join(' - ');
+    if(stringify(log.meta)==="{}")
```
Comment (lazywithclass) [inline index.js]: What about `_.isEmpty`?
Also could you please expand on why `meta` is necessary in addition to `msg`? (I am not sure I got that so I'm missing something here).

File: index.js [hunk 2]
```diff
@@ -55,11 +60,15 @@ util.inherits(WinstonCloudWatch, winston.Transport);
 
 WinstonCloudWatch.prototype.log = function(level, msg, meta, callback) {
   var log = { level: level, msg: msg, meta: meta };
-  this.add(log);
+  if(msg.length!==0){
```
Comment (lazywithclass) [inline index.js]: Same as above, I would use `_.isEmpty`, what do you think?

File: index.js [hunk 3]
```diff
@@ -55,11 +60,15 @@ util.inherits(WinstonCloudWatch, winston.Transport);
...
+    this.add(log);
+  }
 
   if (!/^uncaughtException: /.test(msg)) {
     // do not wait, just return right away
-    return callback(null, true);
+    if(msg.length!==0) {
```
Comment (lazywithclass) [inline index.js]: `_.isEmpty`?

File: index.js [hunk 4]
```diff
@@ -72,10 +81,13 @@ WinstonCloudWatch.prototype.log = function(level, msg, meta, callback) {
 WinstonCloudWatch.prototype.add = function(log) {
   var self = this;
 
-  self.logEvents.push({
-    message: self.formatMessage(log),
-    timestamp: new Date().getTime()
-  });
+  if(log.msg.length!==0)
```
Comment (lazywithclass) [inline index.js]: `_.isEmpty`?

File: index.js [hunk 5]
```diff
@@ -96,13 +108,16 @@ WinstonCloudWatch.prototype.submit = function(callback) {
...
     self.logStreamName() : self.logStreamName;
 
-  cloudWatchIntegration.upload(
-    self.cloudwatchlogs,
-    groupName,
-    streamName,
-    self.logEvents,
-    callback
-  );
+  if(self.logEvents.length!==0)
```
Comment (lazywithclass) [inline index.js]: `_.isEmpty`?

File: index.js (hunk 2 above)
Comment (FlorianBouron) [inline index.js]: Yes very good idea 👍

File: index.js (hunk 3 above)
Comment (FlorianBouron) [inline index.js]: yes :)

File: index.js (hunk 4 above)
Comment (FlorianBouron) [inline index.js]: yes :)

File: index.js (hunk 5 above)
Comment (FlorianBouron) [inline index.js]: yes :)

File: index.js (hunk 1 above)
Comment (FlorianBouron) [inline index.js]: Yes `_.isEmpty` is perfect :)

Here, I am checking if meta contains something because, in the previous code whatever if we had something we sent it to CloudWatch Log. The "issue" was if we didn't have any meta data we have something like this in our log :
`level - message - {}`
This empty bracket wasn't beautiful, that's why I am checking if the meta data.

Comment (lazywithclass) [inline index.js]: Ah I see, makes perfect sense.

Comment (FlorianBouron) [inline index.js]: Perfect, do you have any other questions?

Comment (lazywithclass) [inline index.js]: No seems good now, I will do the changes and fix the tests later on this week, unless you're willing to have a go there.

Comment (FlorianBouron) [inline index.js]: Yes please, I'm waiting for it in order to pull it on my server :)

Comment (lazywithclass) [inline index.js]: 👍

**Comment by lazywithclass** [comment]
I pulled your PR and applied the changes we discussed, I just need to cover those with tests now. Had a busy weekend :b

**Comment by lazywithclass** [comment]
@FlorianBouron there are some edge cases that are not currently covered, for which the `callback` would never be called, I am going through those and make sure everything works as expected.

Do you have a staging server on which we could test a pre-release version? So that we make sure everything is ok? Would you be available to do this?

**Comment by FlorianBouron** [comment]
@lazywithclass sorry I don't have a staging server.

**Comment by lazywithclass** [comment]
@FlorianBouron I did some tests over the weekend, it seems ok. I will merge in the upcoming days.

**Comment by lazywithclass** [comment]
@FlorianBouron I've opened a PR (
https://github.com/lazywithclass/winston-cloudwatch/pull/52) which I think is ready for going live, the version will be 1.8.0

Could you please have a look if it contains everything you wanted? I've taken the code from your branch and added tests and a bit of documentation.

I am closing here.
